File: apps/mobile/lib/src/features/history/data/session_review.dart
import 'dart:convert';

import 'package:crypto/crypto.dart';

import 'session_history_models.dart';

class SessionReview {
  const SessionReview({
    required this.summary,
    this.title,
    this.decisions = const <String>[],
    this.actionItems = const <SessionActionItem>[],
    this.keyFacts = const <SessionKeyFact>[],
    this.risks = const <String>[],
    this.openQuestions = const <String>[],
    required this.highlights,
    required this.terms,
    this.evidenceSegmentIds = const <String>[],
    this.generationKind,
    this.sourceFingerprint,
  });

  final String summary;
  final String? title;
  final List<String> decisions;
  final List<SessionActionItem> actionItems;
  final List<SessionKeyFact> keyFacts;
  final List<String> risks;
  final List<String> openQuestions;
  final List<SessionHighlight> highlights;
  final List<SessionTermSuggestion> terms;
  final List<String> evidenceSegmentIds;
  final String? generationKind;
  final String? sourceFingerprint;
}

class SessionActionItem {
  const SessionActionItem({
    required this.text,
    this.completed = false,
    this.owner,
    this.dueDate,
    this.evidenceSegmentIds = const <String>[],
  });

  final String text;
  final bool completed;
  final String? owner;
  final String? dueDate;
  final List<String> evidenceSegmentIds;
}

class SessionKeyFact {
  const SessionKeyFact({
    required this.type,
    required this.text,
    this.evidenceSegmentIds = const <String>[],
  });

  final String type;
  final String text;
  final List<String> evidenceSegmentIds;
}

class SessionHighlight {
  const SessionHighlight({
    required this.type,
    required this.text,
  });

  final String type;
  final String text;
}

class SessionTermSuggestion {
  const SessionTermSuggestion({
    required this.sourceText,
    required this.translatedText,
  });

  final String sourceText;
  final String translatedText;
}

SessionReview buildSessionReview(SessionDetail detail) {
  final serverReview = _serverReview(detail.reviewJson);
  if (serverReview != null) return serverReview;
  final textPairs = detail.segments
      .where((segment) =>
          segment.sourceText.trim().isNotEmpty ||
          segment.translatedText.trim().isNotEmpty)
      .toList(growable: false);
  return _deviceRuleReview(detail, textPairs);
}

Map<String, Object?> buildDeviceRuleReviewJson(
  SessionDetail detail,
  DateTime generatedAt,
) {
  final review = _deviceRuleReview(detail, _textPairs(detail));
  return <String, Object?>{
    'provider': 'local',
    'promptVersion': 'session_review_device_rules_v1',
    'generatedAt': generatedAt.toUtc().toIso8601String(),
    'generationKind': 'device_rules',
    'sourceFingerprint': deviceRuleReviewSourceFingerprint(detail),
    if (review.title != null) 'title': review.title,
    'summary': review.summary,
    'decisions': review.decisions,
    'actionItems': review.actionItems
        .map((item) => <String, Object?>{
              'text': item.text,
              'completed': item.completed,
              if (item.owner != null) 'owner': item.owner,
              if (item.dueDate != null) 'dueDate': item.dueDate,
              'evidenceSegmentIds': item.evidenceSegmentIds,
            })
        .toList(growable: false),
    'keyFacts': review.keyFacts
        .map((item) => <String, Object?>{
              'type': item.type,
              'text': item.text,
              'evidenceSegmentIds': item.evidenceSegmentIds,
            })
        .toList(growable: false),
    'risks': review.risks,
    'openQuestions': review.openQuestions,
    'highlights': review.highlights
        .map((item) => <String, Object?>{'type': item.type, 'text': item.text})
        .toList(growable: false),
    'terms': review.terms
        .map((item) => <String, Object?>{
              'sourceText': item.sourceText,
              'translatedText': item.translatedText,
            })
        .toList(growable: false),
    'evidenceSegmentIds': review.evidenceSegmentIds,
  };
}

bool isCurrentDeviceRuleReview(SessionDetail detail) {
  final review = detail.reviewJson;
  return review?['generationKind'] == 'device_rules' &&
      review?['sourceFingerprint'] == deviceRuleReviewSourceFingerprint(detail);
}

Map<String, String> confirmedTermIdsForSession(SessionDetail detail) {
  final values = detail.reviewJson?['confirmedTerms'] as List<dynamic>? ??
      const <dynamic>[];
  final result = <String, String>{};
  for (final value in values) {
    if (value is! Map) continue;
    final id = value['id'] as String?;
    final source = value['sourceText'] as String?;
    final translated = value['translatedText'] as String?;
    if (id == null ||
        source == null ||
        translated == null ||
        value['status'] != 'active') {
      continue;
    }
    result[sessionReviewTermKey(source, translated)] = id;
  }
  return result;
}

String sessionReviewTermKey(String sourceText, String translatedText) {
  return '${sourceText.trim().toLowerCase()}\n'
      '${translatedText.trim().toLowerCase()}';
}

String deviceRuleReviewSourceFingerprint(SessionDetail detail) {
  final source = _textPairs(detail)
      .map((segment) => <String, Object?>{
            'id': segment.id,
            'turnId': segment.turnId,
            'revision': segment.revision,
            'rawText': segment.rawText,
            'optimizedText': segment.optimizedText,
            'sourceText': segment.sourceText,
            'translatedText': segment.translatedText,
            'sourceLanguage': segment.sourceLanguage,
            'targetLanguage': segment.targetLanguage,
            'speakerId': segment.speaker?.speakerId,
            'startMs': segment.timing?.startMs,
            'endMs': segment.timing?.endMs,
          })
      .toList(growable: false);
  return sha256.convert(utf8.encode(jsonEncode(source))).toString();
}

SessionReview _deviceRuleReview(
  SessionDetail detail,
  List<SessionSegment> textPairs,
) {
  final highlights = _highlights(textPairs);
  return SessionReview(
    title: _title(textPairs),
    summary: _summary(textPairs),
    actionItems: _actionItems(textPairs, highlights),
    keyFacts: _keyFacts(textPairs, highlights),
    highlights: highlights,
    terms: _terms(textPairs),
    evidenceSegmentIds: textPairs.take(5).map((segment) => segment.id).toList(),
    generationKind: 'device_rules',
    sourceFingerprint: deviceRuleReviewSourceFingerprint(detail),
  );
}

List<SessionSegment> _textPairs(SessionDetail detail) => detail.segments
    .where((segment) =>
        segment.sourceText.trim().isNotEmpty ||
        segment.translatedText.trim().isNotEmpty)
    .toList(growable: false);

SessionReview? _serverReview(Map<String, Object?>? json) {
  if (json == null) return null;
  final title = (json['title'] as String? ?? '').trim();
  final summary = (json['summary'] as String? ?? '').trim();
  final decisions = _stringList(json['decisions']);
  final risks = _stringList(json['risks']);
  final openQuestions = _stringList(json['openQuestions']);
  final actionItems = (json['actionItems'] as List<dynamic>? ?? const [])
      .whereType<Map<String, Object?>>()
      .map((item) => SessionActionItem(
            text: item['text'] as String? ?? '',
            completed: item['completed'] as bool? ?? false,
            owner: item['owner'] as String?,
            dueDate: item['dueDate'] as String?,
            evidenceSegmentIds: _stringList(item['evidenceSegmentIds']),
          ))
      .where((item) => item.text.trim().isNotEmpty)
      .toList(growable: false);
  final keyFacts = (json['keyFacts'] as List<dynamic>? ?? const [])
      .whereType<Map<String, Object?>>()
      .map((item) => SessionKeyFact(
            type: item['type'] as String? ?? 'custom',
            text: item['text'] as String? ?? '',
            evidenceSegmentIds: _stringList(item['evidenceSegmentIds']),
          ))
      .where((item) => item.text.trim().isNotEmpty)
      .toList(growable: false);
  final highlights = (json['highlights'] as List<dynamic>? ?? const [])
      .whereType<Map<String, Object?>>()
      .map((item) => SessionHighlight(
            type: item['type'] as String? ?? 'custom',
            text: item['text'] as String? ?? '',
          ))
      .where((item) => item.text.trim().isNotEmpty)
      .toList(growable: false);
  final terms = (json['terms'] as List<dynamic>? ?? const [])
      .whereType<Map<String, Object?>>()
      .map((item) => SessionTermSuggestion(
            sourceText: item['sourceText'] as String? ?? '',
            translatedText: item['translatedText'] as String? ?? '',
          ))
      .where((item) =>
          item.sourceText.trim().isNotEmpty &&
          item.translatedText.trim().isNotEmpty)
      .toList(growable: false);
  if (summary.isEmpty &&
      highlights.isEmpty &&
      terms.isEmpty &&
      decisions.isEmpty &&
      actionItems.isEmpty &&
      keyFacts.isEmpty) {
    return null;
  }
  return SessionReview(
    title: title.isEmpty ? null : title,
    summary: summary,
    decisions: decisions,
    actionItems: actionItems,
    keyFacts: keyFacts,
    risks: risks,
    openQuestions: openQuestions,
    highlights: highlights,
    terms: terms,
    evidenceSegmentIds: _stringList(json['evidenceSegmentIds']),
    generationKind: json['generationKind'] as String?,
    sourceFingerprint: json['sourceFingerprint'] as String?,
  );
}

String? _title(List<SessionSegment> segments) {
  for (final segment in segments) {
    final text = segment.sourceText.trim().isNotEmpty
        ? segment.sourceText.trim()
        : segment.translatedText.trim();
    if (text.isNotEmpty) {
      return text.length <= 40 ? text : text.substring(0, 40);
    }
  }
  return null;
}

List<SessionActionItem> _actionItems(
  List<SessionSegment> segments,
  List<SessionHighlight> highlights,
) =>
    highlights
        .where((item) => item.type == 'todo')
        .map((item) => SessionActionItem(
              text: item.text,
              evidenceSegmentIds: _evidenceForText(segments, item.text),
            ))
        .where((item) => item.evidenceSegmentIds.isNotEmpty)
        .toList(growable: false);

List<SessionKeyFact> _keyFacts(
  List<SessionSegment> segments,
  List<SessionHighlight> highlights,
) =>
    highlights
        .where((item) => item.type != 'todo' && item.type != 'custom')
        .map((item) => SessionKeyFact(
              type: item.type,
              text: item.text,
              evidenceSegmentIds: _evidenceForText(segments, item.text),
            ))
        .where((item) => item.evidenceSegmentIds.isNotEmpty)
        .toList(growable: false);

List<String> _evidenceForText(List<SessionSegment> segments, String text) =>
    segments
        .where((segment) {
          final source = segment.sourceText.trim();
          final translated = segment.translatedText.trim();
          return (source.isNotEmpty && text.contains(source)) ||
              (translated.isNotEmpty && text.contains(translated));
        })
        .map((segment) => segment.id)
        .take(3)
        .toList(growable: false);

List<String> _stringList(Object? value) {
  return (value as List<dynamic>? ?? const [])
      .whereType<String>()
      .map((item) => item.trim())
      .where((item) => item.isNotEmpty)
      .toList(growable: false);
}

String _summary(List<SessionSegment> segments) {
  if (segments.isEmpty) return '';
  final firstLines = segments.take(3).map((segment) {
    final translated = segment.translatedText.trim();
    if (translated.isNotEmpty) return translated;
    return segment.sourceText.trim();
  }).where((text) => text.isNotEmpty);
  return firstLines.join('\n');
}

List<SessionHighlight> _highlights(List<SessionSegment> segments) {
  final highlights = <SessionHighlight>[];
  for (final segment in segments) {
    final text = _combinedText(segment);
    if (text.isEmpty) continue;
    if (_hasTime(text)) {
      highlights.add(SessionHighlight(type: 'time', text: text));
    } else if (_hasMoney(text)) {
      highlights.add(SessionHighlight(type: 'money', text: text));
    } else if (_hasAction(text)) {
      highlights.add(SessionHighlight(type: 'todo', text: text));
    } else if (_hasLocation(text)) {
      highlights.add(SessionHighlight(type: 'location', text: text));
    }
    if (highlights.length >= 8) break;
  }
  return highlights;
}

List<SessionTermSuggestion> _terms(List<SessionSegment> segments) {
  final seen = <String>{};
  final terms = <SessionTermSuggestion>[];
  for (final segment in segments) {
    final source = segment.sourceText.trim();
    final translated = segment.translatedText.trim();
    if (source.isEmpty || translated.isEmpty) continue;
    if (!_looksLikeTerm(source)) continue;
    final key = source.toLowerCase();
    if (!seen.add(key)) continue;
    terms.add(SessionTermSuggestion(
      sourceText: source,
      translatedText: translated,
    ));
    if (terms.length >= 12) break;
  }
  return terms;
}

String _combinedText(SessionSegment segment) {
  final source = segment.sourceText.trim();
  final translated = segment.translatedText.trim();
  if (source.isEmpty) return translated;
  if (translated.isEmpty) return source;
  return '$source\n$translated';
}

bool _hasTime(String text) {
  return RegExp(r"(\d{1,2}[:：]\d{2}|上午|下午|今天|明天|o'clock|tomorrow)",
          caseSensitive: false)
      .hasMatch(text);
}

bool _hasMoney(String text) {
  return RegExp(r'([$¥￥]\s?\d+|\d+\s?(元|美元|yuan|dollars?))',
          caseSensitive: false)
      .hasMatch(text);
}

bool _hasAction(String text) {
  return RegExp(r'(需要|请|安排|确认|发送|整理|need|please|confirm|send|arrange)',
          caseSensitive: false)
      .hasMatch(text);
}

bool _hasLocation(String text) {
  return RegExp(r'(地址|地点|会议室|room|address|location)', caseSensitive: false)
      .hasMatch(text);
}

bool _looksLikeTerm(String source) {
  if (source.length < 2 || source.length > 24) return false;
  if (RegExp(r'[，。！？,.!?]').hasMatch(source)) return false;
  return RegExp(r'[\u4e00-\u9fffA-Za-z]').hasMatch(source);
}

File: services/api-server/src/modules/sessions/session-review.ts
import {
  createLlmProvider,
  loadLlmConfig,
  type SessionReviewResult,
} from "@translation/llm";
import type {
  SessionReviewHighlightDto,
  SessionReviewResponse,
  SessionReviewTermDto,
} from "@translation/contracts";
import type { SessionRecord } from "./session-record.js";
import {
  recordRemoteReviewFailure,
  recordRemoteReviewSuccess,
} from "./session-review-provider-status.js";
export {
  resetSessionReviewProviderRuntimeStatusForTest,
  sessionReviewProviderStatus,
} from "./session-review-provider-status.js";

const maxRemoteReviewSegments = 40;
const maxReviewSourceCharacters = 180;
const maxReviewTranslationCharacters = 240;

export interface SessionReviewOptions {
  now?: Date;
  fetchFn?: typeof fetch;
}

export async function generateSessionReview(
  session: SessionRecord,
  options: SessionReviewOptions = {},
): Promise<SessionReviewResponse> {
  const config = loadLlmConfig();
  if (config.reviewEnabled && config.provider !== "off") {
    const provider = createLlmProvider(config, options.fetchFn);
    const health = await provider.healthCheck();
    if (health.status !== "ready") {
      const reason =
        health.issues.join("; ") || "LLM review provider unavailable";
      recordRemoteReviewFailure(config, reason, options.now);
      return localSessionReviewFallback(session, options.now, reason);
    }
    try {
      const review = await provider.generateReview({
        sessionId: session.id,
        segments: remoteReviewSegments(session),
      });
      recordRemoteReviewSuccess(config, options.now);
      return toSessionReviewResponse(review, options.now);
    } catch (error) {
      const reason = error instanceof Error ? error.message : String(error);
      recordRemoteReviewFailure(config, reason, options.now);
      return localSessionReviewFallback(session, options.now, reason);
    }
  }
  return localSessionReview(session, options.now);
}

function remoteReviewSegments(session: SessionRecord) {
  return sampleSegmentsForRemoteReview(session.segments)
    .map((segment) => ({
      id: segment.id,
      rawText: compactReviewText(segment.rawText, maxReviewSourceCharacters),
      optimizedText: compactReviewText(segment.optimizedText, maxReviewSourceCharacters),
      sourceText: compactReviewText(segment.sourceText, maxReviewSourceCharacters),
      translatedText: compactReviewText(segment.translatedText, maxReviewTranslationCharacters),
      speaker: segment.speaker?.displayName ?? segment.speaker?.speakerId,
      startedAtMs: segment.timing?.startMs,
      endedAtMs: segment.timing?.endMs,
      endpointReason: segment.vadContext?.endpointReason,
    }));
}

function sampleSegmentsForRemoteReview(segments: SessionRecord["segments"]) {
  const meaningful = segments.filter(
    (segment) =>
      segment.rawText?.trim() ||
      segment.optimizedText?.trim() ||
      segment.sourceText?.trim() ||
      segment.translatedText?.trim(),
  );
  if (meaningful.length <= maxRemoteReviewSegments) return meaningful;

  const result: SessionRecord["segments"] = [];
  const seen = new Set<string>();
  const step = (meaningful.length - 1) / (maxRemoteReviewSegments - 1);
  for (let index = 0; index < maxRemoteReviewSegments; index += 1) {
    const segment = meaningful[Math.round(index * step)];
    if (!segment || seen.has(segment.id)) continue;
    seen.add(segment.id);
    result.push(segment);
  }
  return result;
}

function compactReviewText(value: string | undefined, maxLength: number) {
  const text = value?.replace(/\s+/g, " ").trim();
  if (!text) return undefined;
  if (text.length <= maxLength) return text;
  return `${text.slice(0, maxLength - 3).trim()}...`;
}

export function localSessionReview(
  session: SessionRecord,
  now = new Date(),
): SessionReviewResponse {
  const segments = textSegments(session);
  return {
    provider: "local",
    promptVersion: "session_review_local_v2",
    generatedAt: now.toISOString(),
    title: title(segments),
    summary: summary(segments),
    decisions: [],
    actionItems: actionItems(segments),
    keyFacts: keyFacts(segments),
    risks: [],
    openQuestions: [],
    highlights: highlights(segments),
    terms: terms(segments),
    evidenceSegmentIds: segments.slice(0, 5).map((segment) => segment.id),
  };
}

function localSessionReviewFallback(
  session: SessionRecord,
  now: Date | undefined,
  reason: string,
) {
  const review = localSessionReview(session, now);
  return {
    ...review,
    risks: [
      ...(review.risks ?? []),
      `LLM 纪要暂不可用，已使用本地纪要：${sanitizeFallbackReason(reason)}`,
    ],
  };
}

function sanitizeFallbackReason(reason: string) {
  return reason.replace(/\s+/g, " ").slice(0, 180);
}

function toSessionReviewResponse(
  review: SessionReviewResult,
  now = new Date(),
): SessionReviewResponse {
  return {
    provider: review.provider,
    ...(review.model ? { model: review.model } : {}),
    ...(review.promptVersion ? { promptVersion: review.promptVersion } : {}),
    generatedAt: now.toISOString(),
    ...(review.title ? { title: review.title } : {}),
    summary: review.summary,
    decisions: review.decisions,
    actionItems: review.actionItems,
    keyFacts: review.keyFacts,
    risks: review.risks,
    openQuestions: review.openQuestions,
    highlights: review.highlights,
    terms: review.terms,
    evidenceSegmentIds: review.evidenceSegmentIds,
  };
}

function textSegments(session: SessionRecord) {
  return session.segments
    .map((segment) => ({
      id: segment.id,
      sourceText: (
        segment.optimizedText ??
        segment.sourceText ??
        segment.rawText ??
        ""
      ).trim(),
      rawText: (segment.rawText ?? "").trim(),
      translatedText: segment.translatedText.trim(),
      speaker: segment.speaker?.displayName ?? segment.speaker?.speakerId,
    }))
    .filter((segment) => segment.sourceText || segment.translatedText);
}

function title(segments: ReturnType<typeof textSegments>) {
  const first = segments.find((segment) => segment.sourceText || segment.translatedText);
  return (first?.sourceText || first?.translatedText || "会话纪要").slice(0, 40);
}

function summary(segments: ReturnType<typeof textSegments>) {
  return segments
    .slice(0, 3)
    .map((segment) => {
      const text = segment.translatedText || segment.sourceText;
      return segment.speaker ? `${segment.speaker}：${text}` : text;
    })
    .filter(Boolean)
    .join("\n");
}

function highlights(segments: ReturnType<typeof textSegments>) {
  const result: SessionReviewHighlightDto[] = [];
  for (const segment of segments) {
    const combined = [segment.sourceText, segment.translatedText].filter(Boolean).join("\n");
    const type = classifyHighlight(combined);
    if (!type) continue;
    result.push({ type, text: combined.slice(0, 300) });
    if (result.length >= 8) break;
  }
  return result;
}

function terms(segments: ReturnType<typeof textSegments>) {
  const seen = new Set<string>();
  const result: SessionReviewTermDto[] = [];
  for (const segment of segments) {
    if (!segment.sourceText || !segment.translatedText) continue;
    if (!looksLikeTerm(segment.sourceText)) continue;
    if (isSpelledIdentifierTerm(segment.sourceText, segment.translatedText)) continue;
    const key = segment.sourceText.toLowerCase();
    if (!seen.add(key)) continue;
    result.push({
      sourceText: segment.sourceText,
      translatedText: segment.translatedText,
    });
    if (result.length >= 12) break;
  }
  return result;
}

function isSpelledIdentifierTerm(sourceText: string, translatedText: string) {
  const source = sourceText.trim();
  const translated = translatedText.trim();
  if (source.toLowerCase() !== translated.toLowerCase()) return false;
  const tokens = source.split(/[\s,，、-]+/u).filter(Boolean);
  if (tokens.length === 0 || tokens.length > 6) return false;
  return tokens.every((token) => /^[A-Za-z0-9]{1,6}$/u.test(token)) &&
    tokens.filter((token) => /^[A-Za-z]$/u.test(token)).length >= Math.min(2, tokens.length);
}

function actionItems(segments: ReturnType<typeof textSegments>) {
  return highlights(segments)
    .filter((item) => item.type === "todo")
    .map((item) => ({
      text: item.text,
      evidenceSegmentIds: segments
        .filter((segment) => item.text.includes(segment.sourceText))
        .map((segment) => segment.id)
        .slice(0, 3),
    }));
}

function keyFacts(segments: ReturnType<typeof textSegments>) {
  return highlights(segments)
    .filter((item) => item.type !== "todo" && item.type !== "custom")
    .map((item) => ({
      type: item.type === "todo" ? "custom" : item.type,
      text: item.text,
      evidenceSegmentIds: segments
        .filter((segment) => item.text.includes(segment.sourceText))
        .map((segment) => segment.id)
        .slice(0, 3),
    }));
}

function classifyHighlight(text: string): SessionReviewHighlightDto["type"] | null {
  if (/(\d{1,2}[:：]\d{2}|上午|下午|今天|明天|o'clock|tomorrow)/i.test(text)) return "time";
  if (/([$¥￥]\s?\d+|\d+\s?(元|美元|yuan|dollars?))/i.test(text)) return "money";
  if (/(需要|请|安排|确认|发送|整理|need|please|confirm|send|arrange)/i.test(text)) return "todo";
  if (/(地址|地点|会议室|room|address|location)/i.test(text)) return "location";
  if (/(\+?\d[\d -]{5,}\d)/.test(text)) return "number";
  return null;
}

function looksLikeTerm(source: string) {
  if (source.length < 2 || source.length > 24) return false;
  if (/[，。！？,.!?]/.test(source)) return false;
  return /[\u4e00-\u9fffA-Za-z]/.test(source);
}
